Miscellaneous additions

Cooliokid956 opened the pull request:
- Fixed script file path shortening for Windows in console errors (what I could at least, the one left unfixed is preprocessed and I can't mess with that)
- Adjusted audio sanity check messages to sound more natural
- 9 new functions added:

## `audio_stream_set_loop_points(audio, start, end)`
Sets an audio stream's loop points in samples, no longer requiring a hacky method to loop streams in mods

## `get_active_mod()`
Returns the mod currently being processed, can be used by mods with APIs to determine what they're being used by and react accordingly

## `get_area_update_counter()`
Returns `gAreaUpdateCounter`, can be used to determine various in-game statistics, as well as if the game is being updated

## Exposed several object pointers
`get_current_object()` to retrieve `gCurrentObject`
`get_dialog_object()` to retrieve `gContinueDialogFunctionObject`
`(get|set)_cutscene_focus()` to retrieve/assign `gCutsceneFocus`
`(get|set)_secondary_camera_focus()` to retrieve/assign `gSecondCameraFocus`

Sunketchupm commented:
`gObjectCutscebeDone` being exposed when

Cooliokid956 commented:
One extra change I'd thought about a while ago but just remembered:

Allow ALL characters beyond the Unit Separator character for use with mod storage (with an exception for `[`, `]`, and `=` in keys)

Cooliokid956 commented:
> `gObjectCutsceneDone` being exposed when

You might be hallucinating
![image](https://github.com/user-attachments/assets/33deb59e-1454-4192-9d98-42b77cbde1fb)

Sunketchupm commented:
Might be a similar variable name, since I remember that it exists within `sparkle_star_spawn.inc.c` (or some similar file name)

PeachyPeachSM64 commented on src/pc/mods/mod_storage.cpp:
```diff
@@ -43,11 +43,11 @@ static void strdelete(char* string, const char* substr) {
 
     while (*buffer != '\0') {
-        if ((*buffer >= 'a' && *buffer <= 'z') || (*buffer >= 'A' && *buffer <= 'Z') || (*buffer >= '0' && *buffer <= '9') || *buffer == '_' || *buffer == '.' || *buffer == '-') {
+        if (*buffer > '' && !(isKey && (*buffer == '[' || *buffer == ']' || *buffer == '='))) {
```
how is this compiling? GCC doesn't even allow it
even if it does, that sounds like undefined behavior
![image](https://github.com/user-attachments/assets/e312635a-4186-4f65-b468-44aa85660e0e)

Cooliokid956 replied:
there should be a `` there, but it seems that wasn't pushed through? i'll replace it with a space check instead

PeachyPeachSM64 replied:
> there should be a � there

no, never use special characters in code, what the hell
thanks git for preventing some war crimes

AgentXLP replied:
The horrors

PeachyPeachSM64 commented on src/pc/lua/utils/smlua_misc_utils.h:
```diff
@@ -41,6 +41,8 @@ struct DateTime {
 /* |description|Gets the current area's networked timer|descriptionEnd| */
 u32 get_network_area_timer(void);
+/* |description|Gets the area update counter incremented when objects are updated|descriptionEnd| */
+u16 get_area_update_counter(void);
```
what's the use for this? why this one and not gGlobalTimer?

Cooliokid956 replied:
`gAreaUpdateCounter` only updates in certain contexts while gGlobalTimer is always ticking

PeachyPeachSM64 replied:
what are those "certain contexts"? how is it useful in lua mods? just curious

AgentXLP replied:
"To determine Mario's true blinking state"

PeachyPeachSM64 commented on src/pc/mods/mod_storage.cpp:
```diff
@@ -148,7 +148,7 @@ C_FIELD bool mod_storage_load_bool(const char* key) {
     if (gLuaActiveMod == NULL) { return false; }
     if (strlen(key) > MAX_KEY_VALUE_LENGTH) { return false; }
-    if (!char_valid((char *)key)) { return false; }
+    if (!char_valid((char *)key, true)) { return false; }
```
pointless cast

Cooliokid956 replied:
hey, I'm not the one to `Blame`!

PeachyPeachSM64 commented on src/pc/mods/mod_storage.cpp:
```diff
@@ -164,7 +164,7 @@ C_FIELD bool mod_storage_exists(const char* key) {
     if (gLuaActiveMod == NULL) { return false; }
     if (strlen(key) > MAX_KEY_VALUE_LENGTH) { return false; }
-    if (!char_valid((char *)key)) { return false; }
+    if (!char_valid((char *)key, true)) { return false; }
```
same here

Cooliokid956 replied:
:p

Cooliokid956 commented:
I was asked to expose one more object:
`geo_get_current_object()` to retrieve `gCurGraphNodeObject` cast to a `struct Object*` which somehow works and is used throughout the game, useful in geo asm functions

Isaac0-dev reviewed (changes requested):
looks mostly fine

AgentXLP commented on src/pc/lua/utils/smlua_audio_utils.c:
```diff
@@ -349,88 +346,82 @@ void audio_stream_play(struct ModAudio* audio, bool restart, f32 volume) {
     ma_sound_set_looping(&audio->sound, looping);
 }
 
+void audio_stream_set_loop_points(struct ModAudio* audio, u64 start, u64 end) {
```
Suggestion: loopStart and loopEnd

Cooliokid956 replied:
I did see an error in the autogen file on vsc, but shrugged it off since it loaded just fine ingame (probably not the actual script though)

will apply these changes

Cooliokid956 commented on src/pc/lua/utils/smlua_audio_utils.c:
```diff
@@ -324,18 +324,15 @@ struct ModAudio* audio_stream_load(const char* filename) {
 }
```
fixed

Cooliokid956 commented:
That was something
here's the changelog from the commit description, plus changes not listed

- switched to loopStart/End wording
- switched param types to `s64`
- reverted return curly braces

Additionally, I:
- found two missing spaces after a for and if
- fixed `audio_stream_get_position`, now getting the true position no matter the loop point setup
- adjusted `audio_stream_set_loop_points`'s behavior to accept negative numbers, these will wrap around to the end of the stream (you may set `loopEnd` to `0` to set the loop end to the very end)
- provide a hint for rehabilitating the tempo functions (?)
- added missing `gCurGraphNodeObject` extern

Cooliokid956 commented:
Today I worked on a few more changes for autogen

1. Make array types generate as actual arrays instead of using `Array_`:
![image](https://github.com/user-attachments/assets/6b520ac0-f4db-4d4e-bfcc-5ca8d1a80b32)

2. Overhauled enum constant autogen to allow for enum autocomplete as such:
![image](https://github.com/user-attachments/assets/dddd123a-fe1f-4487-91c8-9249a74054b8)

AgentXLP commented:
It's peak...

Isaac0-dev commented on src/pc/lua/utils/smlua_audio_utils.c:
```diff
@@ -442,9 +438,8 @@ void audio_stream_set_volume(struct ModAudio* audio, f32 volume) {
-//     if (!audio_sanity_check(audio, true, "set stream speed")) {
+//     if (!audio_sanity_check(audio, true, "set stream speed for"))
 //         return;
-//     }
```
braces are still removed here

Cooliokid956 replied:
ok ok

Isaac0-dev reviewed (commented):
`Object[]` instead of `Array_Object`, but not `*Object` instead of `Pointer_Object`?
it doesn't look consistent, and if you went with both it wouldn't look great with the `*Object`.
imo just leave it as `Array_Object`.
